File: myExpenses/src/main/java/org/totschnig/myexpenses/fragment/CsvImportDataFragment.java
package org.totschnig.myexpenses.fragment;

import android.app.ProgressDialog;
import android.graphics.Typeface;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.DisplayMetrics;
import android.util.SparseBooleanArray;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.LinearLayout;
import android.widget.Spinner;
import android.widget.TextView;

import org.apache.commons.csv.CSVRecord;
import org.json.JSONException;
import org.json.JSONObject;
import org.totschnig.myexpenses.R;
import org.totschnig.myexpenses.activity.CsvImportActivity;
import org.totschnig.myexpenses.activity.ProtectedFragmentActivity;
import org.totschnig.myexpenses.dialog.ConfirmationDialogFragment;
import org.totschnig.myexpenses.dialog.ProgressDialogFragment;
import org.totschnig.myexpenses.export.qif.QifDateFormat;
import org.totschnig.myexpenses.model.AccountType;
import org.totschnig.myexpenses.model.CurrencyUnit;
import org.totschnig.myexpenses.preference.PrefKey;
import org.totschnig.myexpenses.task.TaskExecutionFragment;
import org.totschnig.myexpenses.util.SparseBooleanArrayParcelable;
import org.totschnig.myexpenses.util.UiUtils;
import org.totschnig.myexpenses.util.Utils;
import org.totschnig.myexpenses.util.crashreporting.CrashHandler;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import timber.log.Timber;

import static org.totschnig.myexpenses.activity.ProtectedFragmentActivity.ASYNC_TAG;
import static org.totschnig.myexpenses.activity.ProtectedFragmentActivity.PROGRESS_TAG;

/**
 * Created by privat on 30.06.15.
 */
public class CsvImportDataFragment extends Fragment {
  public static final String KEY_DATASET = "DATASET";
  public static final String KEY_DISCARDED_ROWS = "DISCARDED_ROWS";
  public static final String KEY_COLUMN_TO_FIELD = "COLUMN_TO_FIELD";
  public static final String KEY_FIELD_TO_COLUMN = "FIELD_TO_COLUMN";
  public static final String KEY_FIRST_LINE_IS_HEADER = "FIRST_LINE_IS_HEADER";

  public static final int CELL_MIN_WIDTH = 100;
  public static final int CHECKBOX_COLUMN_WIDTH = 60;
  public static final int CELL_MARGIN = 5;
  private RecyclerView mRecyclerView;
  private LinearLayout mHeaderLine;
  private RecyclerView.Adapter mAdapter;
  private RecyclerView.LayoutManager mLayoutManager;
  private ArrayList<CSVRecord> mDataset;
  private SparseBooleanArrayParcelable discardedRows;

  private ArrayAdapter<Integer> mFieldAdapter;
  private LinearLayout.LayoutParams cellParams, cbParams;
  private boolean firstLineIsHeader;

  private int nrOfColumns;

  public static final String FIELD_KEY_DISCARD = "DISCARD";
  public static final String FIELD_KEY_AMOUNT = "AMOUNT";
  public static final String FIELD_KEY_EXPENSE = "EXPENSE";
  public static final String FIELD_KEY_INCOME = "INCOME";
  public static final String FIELD_KEY_DATE = "DATE";
  public static final String FIELD_KEY_PAYEE = "PAYEE";
  public static final String FIELD_KEY_COMMENT = "COMMENT";
  public static final String FIELD_KEY_CATEGORY = "CATEGORY";
  public static final String FIELD_KEY_SUBCATEGORY = "SUBACTEGORY";
  public static final String FIELD_KEY_METHOD = "METHOD";
  public static final String FIELD_KEY_STATUS = "STATUS";
  public static final String FIELD_KEY_NUMBER = "NUMBER";
  public static final String FIELD_KEY_SPLIT = "SPLIT";


  private final String[] fieldKeys = new String[]{
      FIELD_KEY_DISCARD, FIELD_KEY_AMOUNT, FIELD_KEY_EXPENSE, FIELD_KEY_INCOME,
      FIELD_KEY_DATE, FIELD_KEY_PAYEE, FIELD_KEY_COMMENT, FIELD_KEY_CATEGORY,
      FIELD_KEY_SUBCATEGORY, FIELD_KEY_METHOD, FIELD_KEY_STATUS, FIELD_KEY_NUMBER,
      FIELD_KEY_SPLIT
  };
  private final Integer[] fields = new Integer[]{
      R.string.cvs_import_discard,
      R.string.amount,
      R.string.expense,
      R.string.income,
      R.string.date,
      R.string.payer_or_payee,
      R.string.comment,
      R.string.category,
      R.string.subcategory,
      R.string.method,
      R.string.status,
      R.string.reference_number,
      R.string.split_transaction
  };
  private JSONObject header2FieldMap;
  private float windowWidth;

  public static CsvImportDataFragment newInstance() {
    return new CsvImportDataFragment();
  }

  @Override
  public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setHasOptionsMenu(true);
  }

  @Override
  public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {

    DisplayMetrics displayMetrics = getResources().getDisplayMetrics();

    windowWidth = displayMetrics.widthPixels / displayMetrics.density;

    String header2FieldMapJson = PrefKey.CSV_IMPORT_HEADER_TO_FIELD_MAP.getString(null);
    if (header2FieldMapJson != null) {
      try {
        header2FieldMap = new JSONObject(header2FieldMapJson);
      } catch (JSONException e) {
        header2FieldMap = new JSONObject();
      }
    } else {
      header2FieldMap = new JSONObject();
    }

    cbParams = new LinearLayout.LayoutParams(UiUtils.dp2Px(CHECKBOX_COLUMN_WIDTH, getResources()),
        LinearLayout.LayoutParams.WRAP_CONTENT);
    cbParams.setMargins(CELL_MARGIN, CELL_MARGIN, CELL_MARGIN, CELL_MARGIN);

    mFieldAdapter = new ArrayAdapter<Integer>(
        getActivity(), android.R.layout.simple_spinner_item, fields) {
      @Override
      public View getView(int position, View convertView, ViewGroup parent) {
        TextView tv = (TextView) super.getView(position, convertView, parent);
        tv.setText(getString(fields[position]));
        return tv;
      }

      @Override
      public View getDropDownView(int position, View convertView, ViewGroup parent) {
        TextView tv = (TextView) super.getDropDownView(position, convertView, parent);
        tv.setText(getString(fields[position]));
        return tv;
      }
    };
    mFieldAdapter.setDropDownViewResource(R.layout.support_simple_spinner_dropdown_item);
    View view = inflater.inflate(R.layout.import_csv_data, container, false);
    mRecyclerView = view.findViewById(R.id.my_recycler_view);
    mHeaderLine = view.findViewById(R.id.header_line);

    // use this setting to improve performance if you know that changes
    // in content do not change the layout size of the RecyclerView
    // http://www.vogella.com/tutorials/AndroidRecyclerView/article.html
    mRecyclerView.setHasFixedSize(true);

    // use a linear layout manager
    mLayoutManager = new LinearLayoutManager(getActivity());
    mRecyclerView.setLayoutManager(mLayoutManager);
    if (savedInstanceState != null) {
      setData((ArrayList<CSVRecord>) savedInstanceState.getSerializable(KEY_DATASET));
      discardedRows = savedInstanceState.getParcelable(KEY_DISCARDED_ROWS);
      firstLineIsHeader = savedInstanceState.getBoolean(KEY_FIRST_LINE_IS_HEADER);
    }

    return view;
  }

  public void setData(ArrayList<CSVRecord> data) {
    if (data == null || data.isEmpty()) return;
    mDataset = data;
    int nrOfColumns = mDataset.get(0).size();
    discardedRows = new SparseBooleanArrayParcelable();

    int availableCellWidth =
        (int) ((windowWidth - CHECKBOX_COLUMN_WIDTH - CELL_MARGIN * (nrOfColumns + 2)) / nrOfColumns);
    int cellWidth, tableWidth;
    if (availableCellWidth > CELL_MIN_WIDTH) {
      cellWidth = availableCellWidth;
      tableWidth = (int) windowWidth;
    } else {
      cellWidth = CELL_MIN_WIDTH;
      tableWidth = CELL_MIN_WIDTH * nrOfColumns + CHECKBOX_COLUMN_WIDTH + CELL_MARGIN * (nrOfColumns + 2);
    }

    cellParams = new LinearLayout.LayoutParams(UiUtils.dp2Px(cellWidth, getResources()),
        LinearLayout.LayoutParams.WRAP_CONTENT);
    cellParams.setMargins(CELL_MARGIN, CELL_MARGIN, CELL_MARGIN, CELL_MARGIN);

    ViewGroup.LayoutParams params = mRecyclerView.getLayoutParams();
    params.width = UiUtils.dp2Px(tableWidth, getResources());
    mRecyclerView.setLayoutParams(params);
    mAdapter = new MyAdapter();
    mRecyclerView.setAdapter(mAdapter);
    //set up header
    mHeaderLine.removeViews(1, mHeaderLine.getChildCount() - 1);
    for (int i = 0; i < nrOfColumns; i++) {
      Spinner cell = new Spinner(getActivity());
      cell.setId(i);
      cell.setAdapter(mFieldAdapter);
      mHeaderLine.addView(cell, cellParams);
    }
  }

  public void setHeader() {
    firstLineIsHeader = true;
    mAdapter.notifyItemChanged(0);
    //automap
    final CSVRecord record = mDataset.get(0);

    outer:
    for (int j = 0; j < record.size(); j++) {
      String headerLabel = Utils.normalize(record.get(j));
      Iterator<String> keys = header2FieldMap.keys();
      while (keys.hasNext()) {
        String storedLabel = keys.next();
        if (storedLabel.equals(headerLabel)) {
          try {
            String fieldKey = header2FieldMap.getString(storedLabel);
            int position = Arrays.asList(fieldKeys).indexOf(fieldKey);
            if (position != -1) {
              ((Spinner) mHeaderLine.getChildAt(j + 1)).setSelection(position);
              continue outer;
            }
          } catch (JSONException e) {
            //ignore
          }
        }
      }
      for (int i = 1 /* 0=Discard ignored  */; i < fields.length; i++) {
        String fieldLabel = Utils.normalize(getString(fields[i]));
        if (fieldLabel.equals(headerLabel)) {
          ((Spinner) mHeaderLine.getChildAt(j + 1)).setSelection(i);
          break;
        }
      }
    }
  }

  private class MyAdapter extends RecyclerView.Adapter<MyAdapter.ViewHolder> implements
      CompoundButton.OnCheckedChangeListener {

    @Override
    public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
      int position = (int) buttonView.getTag();
      Timber.d("%s item at position %d", isChecked ? "Discarding" : "Including", position);
      if (isChecked) {
        discardedRows.put(position, true);
        if (position == 0) {
          Bundle b = new Bundle();
          b.putInt(ConfirmationDialogFragment.KEY_TITLE,
              R.string.dialog_title_information);
          b.putString(
              ConfirmationDialogFragment.KEY_MESSAGE,
              getString(R.string.cvs_import_set_first_line_as_header));
          b.putInt(ConfirmationDialogFragment.KEY_COMMAND_POSITIVE,
              R.id.SET_HEADER_COMMAND);
          ConfirmationDialogFragment.newInstance(b).show(
              getParentFragmentManager(), "SET_HEADER_CONFIRMATION");
        }
      } else {
        discardedRows.delete(position);
        if (position == 0) {
          firstLineIsHeader = false;
        }
      }
      notifyItemChanged(position);
    }

    // Provide a reference to the views for each data item
    // Complex data items may need more than one view per item, and
    // you provide access to all the views for a data item in a view holder
    public class ViewHolder extends RecyclerView.ViewHolder {
      // each data item is just a string in this case
      public LinearLayout row;

      public ViewHolder(LinearLayout v) {
        super(v);
        row = v;
      }
    }

    // Provide a suitable constructor (depends on the kind of dataset)
    MyAdapter() {
      nrOfColumns = mDataset.get(0).size();
    }

    // Create new views (invoked by the layout manager)
    @NonNull
    @Override
    public MyAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent,
                                                   int viewType) {
      // create a new view
      LinearLayout v = new LinearLayout(parent.getContext());
      v.setBackgroundResource(R.drawable.csv_import_row_background);
      TextView cell = new CheckBox(parent.getContext());

      v.addView(cell, cbParams);
      for (int i = 0; i < nrOfColumns; i++) {
        cell = new TextView(parent.getContext());
        cell.setSingleLine();
        cell.setEllipsize(TextUtils.TruncateAt.END);
        cell.setSelected(true);
        cell.setOnClickListener(v1 -> ((ProtectedFragmentActivity) getActivity()).showSnackbar(((TextView) v1).getText()));
        if (viewType == 0) {
          cell.setTypeface(null, Typeface.BOLD);
        }
        v.addView(cell, cellParams);
      }
      // set the view's size, margins, paddings and layout parameters
      ViewHolder vh = new ViewHolder(v);
      return vh;
    }

    // Replace the contents of a view (invoked by the layout manager)
    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, final int position) {
      // - get element from your dataset at this position
      // - replace the contents of the view with that element
      boolean isDiscarded = discardedRows.get(position, false);
      boolean isHeader = position == 0 && firstLineIsHeader;
      holder.row.setActivated(isDiscarded && !isHeader);
      final CSVRecord record = mDataset.get(position);
      for (int i = 0; i < record.size() && i < nrOfColumns; i++) {
        TextView cell = (TextView) holder.row.getChildAt(i + 1);
        cell.setText(record.get(i));
      }
      CheckBox cb = (CheckBox) holder.row.getChildAt(0);
      cb.setTag(position);
      cb.setOnCheckedChangeListener(null);
      cb.setChecked(isDiscarded);
      cb.setOnCheckedChangeListener(this);
    }

    // Return the size of your dataset (invoked by the layout manager)
    @Override
    public int getItemCount() {
      return mDataset.size();
    }

    @Override
    public int getItemViewType(int position) {
      return position == 0 & firstLineIsHeader ? 0 : 1;
    }
  }

  @Override
  public void onSaveInstanceState(@NonNull Bundle outState) {
    super.onSaveInstanceState(outState);
    outState.putSerializable(KEY_DATASET, mDataset);
    outState.putParcelable(KEY_DISCARDED_ROWS, discardedRows);
    outState.putBoolean(KEY_FIRST_LINE_IS_HEADER, firstLineIsHeader);
  }

  @Override
  public void onCreateOptionsMenu(@NonNull Menu menu, MenuInflater inflater) {
    inflater.inflate(R.menu.csv_import, menu);
  }

  @Override
  public boolean onOptionsItemSelected(MenuItem item) {
    if (item.getItemId() == R.id.IMPORT_COMMAND) {
      int[] columnToFieldMap = new int[nrOfColumns];
      final CSVRecord header = mDataset.get(0);
      for (int i = 0; i < nrOfColumns; i++) {
        int position = ((Spinner) mHeaderLine.getChildAt(i + 1)).getSelectedItemPosition();
        columnToFieldMap[i] = fields[position];
        if (firstLineIsHeader) {
          try {
            if (!fieldKeys[position].equals(FIELD_KEY_DISCARD)) {
              header2FieldMap.put(Utils.normalize(header.get(i)), fieldKeys[position]);
            }
          } catch (JSONException e) {
            CrashHandler.report(e);
          }
        }
      }
      if (validateMapping(columnToFieldMap)) {
        PrefKey.CSV_IMPORT_HEADER_TO_FIELD_MAP.putString(header2FieldMap.toString());
        long accountId = ((CsvImportActivity) getActivity()).getAccountId();
        CurrencyUnit currency = ((CsvImportActivity) getActivity()).getCurrency();
        QifDateFormat format = ((CsvImportActivity) getActivity()).getDateFormat();
        AccountType type = ((CsvImportActivity) getActivity()).getAccountType();
        TaskExecutionFragment taskExecutionFragment =
            TaskExecutionFragment.newInstanceCSVImport(
                mDataset, columnToFieldMap, discardedRows, format, accountId, currency, type);
        ProgressDialogFragment progressDialogFragment = ProgressDialogFragment.newInstance(
            getString(R.string.pref_import_title, "CSV"),
            null, ProgressDialog.STYLE_HORIZONTAL, false);
        progressDialogFragment.setMax(mDataset.size() - discardedRows.size());
        getParentFragmentManager()
            .beginTransaction()
            .add(taskExecutionFragment, ASYNC_TAG)
            .add(progressDialogFragment, PROGRESS_TAG)
            .commit();

      }
    }
    return super.onOptionsItemSelected(item);
  }

  /**
   * Check mapping constraints:<br>
   * <ul>
   * <li>No field mapped more than once</li>
   * <li>Subcategory cannot be mapped withoug category</li>
   * <li>One of amount, income or expense must be mapped.</li>
   * </ul>
   *
   * @param columnToFieldMap
   */
  private boolean validateMapping(int[] columnToFieldMap) {
    SparseBooleanArray foundFields = new SparseBooleanArray();
    ProtectedFragmentActivity activity = (ProtectedFragmentActivity) getActivity();
    for (int field : columnToFieldMap) {
      if (field != R.string.cvs_import_discard) {
        if (foundFields.get(field, false)) {
          activity.showSnackbar(getString(R.string.csv_import_field_mapped_more_than_once, getString(field)));
          return false;
        }
        foundFields.put(field, true);
      }
    }
    if (foundFields.get(R.string.subcategory, false) && !foundFields.get(R.string.category, false)) {
      activity.showSnackbar(R.string.csv_import_subcategory_requires_category);
      return false;
    }
    if (!(foundFields.get(R.string.amount, false) ||
        foundFields.get(R.string.expense, false) ||
        foundFields.get(R.string.income, false))) {
      activity.showSnackbar(R.string.csv_import_no_mapping_found_for_amount);
      return false;
    }
    return true;
  }
}

File: myExpenses/src/main/java/org/totschnig/myexpenses/fragment/CsvImportDataFragment.kt
package org.totschnig.myexpenses.fragment

import android.graphics.Typeface
import android.os.Bundle
import android.text.TextUtils
import android.util.SparseBooleanArray
import android.view.Gravity
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.CompoundButton
import android.widget.LinearLayout
import android.widget.LinearLayout.LayoutParams.WRAP_CONTENT
import android.widget.Spinner
import android.widget.TextView
import androidx.appcompat.widget.AppCompatSpinner
import androidx.appcompat.widget.AppCompatTextView
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.updatePadding
import androidx.core.widget.TextViewCompat
import androidx.core.widget.TextViewCompat.AUTO_SIZE_TEXT_TYPE_UNIFORM
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.RecyclerView
import kotlinx.coroutines.launch
import org.apache.commons.csv.CSVRecord
import org.json.JSONException
import org.json.JSONObject
import org.totschnig.myexpenses.MyApplication
import org.totschnig.myexpenses.R
import org.totschnig.myexpenses.activity.CsvImportActivity
import org.totschnig.myexpenses.activity.ProtectedFragmentActivity
import org.totschnig.myexpenses.databinding.ImportCsvDataBinding
import org.totschnig.myexpenses.databinding.ImportCsvDataRowBinding
import org.totschnig.myexpenses.dialog.ConfirmationDialogFragment
import org.totschnig.myexpenses.preference.PrefHandler
import org.totschnig.myexpenses.preference.PrefKey
import org.totschnig.myexpenses.util.Utils
import org.totschnig.myexpenses.util.crashreporting.CrashHandler
import org.totschnig.myexpenses.viewmodel.CsvImportViewModel
import org.totschnig.myexpenses.viewmodel.CsvImportViewModel.Companion.KEY_HEADER_LINE_POSITION
import timber.log.Timber
import javax.inject.Inject

class CsvImportDataFragment : Fragment() {

    private val viewModel: CsvImportViewModel by activityViewModels()

    private var _binding: ImportCsvDataBinding? = null
    private val binding get() = _binding!!
    private lateinit var cellParams: LinearLayout.LayoutParams
    private var nrOfColumns: Int = 0
    private val allFields: List<Pair<Int, String?>> = listOf(
        R.string.discard to null,
        R.string.account to "ACCOUNT",
        R.string.amount to "AMOUNT",
        R.string.expense to "EXPENSE",
        R.string.income to "INCOME",
        R.string.date to "DATE",
        R.string.booking_date to "BOOKING_DATE",
        R.string.value_date to "VALUE_DATE",
        R.string.time to "TIME",
        R.string.payer_or_payee to "PAYEE",
        R.string.notes to "COMMENT",
        R.string.category to "CATEGORY",
        R.string.subcategory to "SUB_CATEGORY",
        R.string.method to "METHOD",
        R.string.status to "STATUS",
        R.string.reference_number to "NUMBER",
        R.string.split_transaction to "SPLIT",
        R.string.tags to "TAGS"
    )

    private lateinit var fields: List<Pair<Int, String?>>
    private lateinit var header2FieldMap: JSONObject
    private var windowWidth = 0
    private var cellMinWidth = 0
    private var checkboxColumnWidth = 0
    private var cellMargin = 0
    private var spinnerRightPadding = 0

    @Inject
    lateinit var prefHandler: PrefHandler

    private val dataSet
        get() = viewModel.dataFlow.value

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        (requireActivity().application as MyApplication).appComponent.inject(this)
        setHasOptionsMenu(true)
        cellMinWidth = resources.getDimensionPixelSize(R.dimen.csv_import_header_text_size) * 8
        checkboxColumnWidth =
            resources.getDimensionPixelSize(R.dimen.csv_import_checkbox_column_width)
        cellMargin = resources.getDimensionPixelSize(R.dimen.csv_import_cell_margin)
        spinnerRightPadding =
            resources.getDimensionPixelSize(R.dimen.csv_import_spinner_right_padding)

        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.dataFlow.collect {
                    setData(it)
                }
            }
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        val displayMetrics = resources.displayMetrics
        windowWidth = displayMetrics.widthPixels
        header2FieldMap = prefHandler.getString(PrefKey.CSV_IMPORT_HEADER_TO_FIELD_MAP, null)?.let {
            try {
                JSONObject(it)
            } catch (_: JSONException) {
                null
            }
        } ?: JSONObject()

        _binding = ImportCsvDataBinding.inflate(inflater, container, false)

        // use this setting to improve performance if you know that changes
        // in content do not change the layout size of the RecyclerView
        // http://www.vogella.com/tutorials/AndroidRecyclerView/article.html
        binding.myRecyclerView.setHasFixedSize(true)

        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        ViewCompat.setOnApplyWindowInsetsListener(binding.scrollViewContainer, { v, insets ->
            val systemBarsAndCutouts = insets.getInsets(
                WindowInsetsCompat.Type.systemBars() or WindowInsetsCompat.Type.displayCutout()
            )

            v.updatePadding(
                left = systemBarsAndCutouts.left,
                right = systemBarsAndCutouts.right
            )

           insets
        })
        ViewCompat.setOnApplyWindowInsetsListener(binding.myRecyclerView, { v, insets ->
            val navigationInsets = insets.getInsets(
                WindowInsetsCompat.Type.navigationBars()
            )

            v.updatePadding(
                bottom = navigationInsets.bottom
            )

            WindowInsetsCompat.CONSUMED
        })
    }

    private fun setData(data: List<CSVRecord>) {
        if (data.isEmpty()) return
        binding.switcher.displayedChild = 1
        val withValueDate = prefHandler.getBoolean(PrefKey.TRANSACTION_WITH_VALUE_DATE, false)
        val withTime =
            !withValueDate && prefHandler.getBoolean(PrefKey.TRANSACTION_WITH_TIME, false)
        fields = allFields.filter {
            when (it.first) {
                R.string.date -> !withValueDate
                R.string.time -> withTime
                R.string.booking_date, R.string.value_date -> withValueDate
                R.string.account -> viewModel.withAccountColumn
                else -> true
            }
        }
        val fieldAdapter = object : ArrayAdapter<Pair<Int, String?>>(
            requireActivity(), R.layout.spinner_item_narrow, 0, fields
        ) {
            override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
                val tv = super.getView(position, convertView, parent) as TextView
                tv.text = getString(fields[position].first)
                return tv
            }

            override fun getDropDownView(
                position: Int,
                convertView: View?,
                parent: ViewGroup,
            ): View {
                val tv = super.getDropDownView(position, convertView, parent) as TextView
                tv.text = getString(fields[position].first)
                return tv
            }
        }.also {
            it.setDropDownViewResource(androidx.appcompat.R.layout.support_simple_spinner_dropdown_item)
        }
        nrOfColumns = data.maxOf { it.size() }
        val availableCellWidth =
            ((windowWidth - checkboxColumnWidth - cellMargin * nrOfColumns * 2) / nrOfColumns)
        val cellWidth: Int
        val tableWidth: Int
        if (availableCellWidth > cellMinWidth) {
            cellWidth = availableCellWidth
            tableWidth = windowWidth
        } else {
            cellWidth = cellMinWidth
            tableWidth =
                cellMinWidth * nrOfColumns + checkboxColumnWidth + cellMargin * nrOfColumns * 2
        }
        cellParams = LinearLayout.LayoutParams(cellWidth, WRAP_CONTENT).apply {
            setMargins(cellMargin, cellMargin, cellMargin, cellMargin)
        }
        with(binding.myRecyclerView) {
            val params = layoutParams
            params.width = tableWidth
            layoutParams = params
            adapter = MyAdapter()
        }

        //set up header
        with(binding.headerLine) {
            removeViews(1, childCount - 1)
            for (i in 0 until nrOfColumns) {
                val cell = AppCompatSpinner(requireContext())
                cell.id = View.generateViewId()
                cell.adapter = fieldAdapter
                cell.setPaddingRelative(0, 0, spinnerRightPadding, 0)
                addView(cell, cellParams)
                viewModel.mapping?.get(i)?.let { cell.setSelection(it) }
            }
        }
    }

    fun setHeader(rowPosition: Int) {
        viewModel.headerLine = rowPosition
        binding.myRecyclerView.adapter?.notifyItemChanged(rowPosition)
        //autoMap
        val record = dataSet[rowPosition]
        outer@ for (j in 0 until record.size()) {
            val headerLabel = Utils.normalize(record[j])
            val keys = header2FieldMap.keys()
            while (keys.hasNext()) {
                val storedLabel = keys.next()
                if (storedLabel == headerLabel) {
                    try {
                        val fieldKey = header2FieldMap.getString(storedLabel)
                        val fieldPosition = fields.indexOfFirst { it.second == fieldKey }
                        if (fieldPosition != -1) {
                            (binding.headerLine.getChildAt(j + 1) as Spinner).setSelection(
                                fieldPosition
                            )
                            continue@outer
                        }
                    } catch (_: JSONException) {
                        //ignore
                    }
                }
            }
            for (i in 1 /* 0=Select ignored  */ until fields.size) {
                val fieldLabel = Utils.normalize(getString(fields[i].first))
                if (fieldLabel == headerLabel) {
                    (binding.headerLine.getChildAt(j + 1) as Spinner).setSelection(i)
                    break
                }
            }
        }
    }

    private inner class MyAdapter : RecyclerView.Adapter<MyAdapter.ViewHolder>(),
        CompoundButton.OnCheckedChangeListener {
        override fun onCheckedChanged(buttonView: CompoundButton, isChecked: Boolean) {
            val position = buttonView.tag as Int
            Timber.d(
                "%s item at position %d",
                if (isChecked) "Selecting" else "Discarding",
                position
            )
            if (isChecked) {
                viewModel.selectRow(position)
                if (position == viewModel.headerLine) {
                    viewModel.headerLine = -1
                }
            } else {
                if (viewModel.headerLine == -1 && position == firstSelectedRow()) {
                    ConfirmationDialogFragment.newInstance(Bundle().apply {
                        putInt(
                            ConfirmationDialogFragment.KEY_TITLE,
                            R.string.information
                        )
                        putString(
                            ConfirmationDialogFragment.KEY_MESSAGE,
                            getString(R.string.cvs_import_set_first_line_as_header)
                        )
                        putInt(
                            ConfirmationDialogFragment.KEY_COMMAND_POSITIVE,
                            R.id.SET_HEADER_COMMAND
                        )
                        putInt(KEY_HEADER_LINE_POSITION, position)
                        putInt(
                            ConfirmationDialogFragment.KEY_POSITIVE_BUTTON_LABEL,
                            R.string.response_yes
                        )
                        putInt(
                            ConfirmationDialogFragment.KEY_NEGATIVE_BUTTON_LABEL,
                            R.string.response_no
                        )
                    }).show(
                        parentFragmentManager, "SET_HEADER_CONFIRMATION"
                    )
                }
                viewModel.unselectRow(position)
            }
            notifyItemChanged(position)
        }

        inner class ViewHolder(val itemBinding: ImportCsvDataRowBinding) :
            RecyclerView.ViewHolder(itemBinding.root)

        override fun onCreateViewHolder(
            parent: ViewGroup,
            viewType: Int,
        ): ViewHolder {
            val itemBinding =
                ImportCsvDataRowBinding.inflate(LayoutInflater.from(parent.context), parent, false)
            repeat(nrOfColumns) {
                val cell = AppCompatTextView(parent.context)
                cell.setSingleLine()
                TextViewCompat.setAutoSizeTextTypeWithDefaults(cell, AUTO_SIZE_TEXT_TYPE_UNIFORM)
                cell.ellipsize = TextUtils.TruncateAt.END
                cell.isSelected = true
                cell.gravity = Gravity.CENTER_HORIZONTAL
                cell.setOnClickListener { v1: View ->
                    (requireActivity() as ProtectedFragmentActivity).showSnackBar(
                        (v1 as TextView).text
                    )
                }
                if (viewType == 0) {
                    cell.setTypeface(null, Typeface.BOLD)
                }
                itemBinding.root.addView(cell, cellParams)
            }
            return ViewHolder(itemBinding)
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            val isSelected = viewModel.isSelected(position)
            val isHeader = position == viewModel.headerLine
            holder.itemView.isActivated = !isSelected && !isHeader
            val record = dataSet[position]
            for (i in 0 until record.size()) {
                val cell = holder.itemBinding.root.getChildAt(i + 1) as TextView
                cell.text = record[i]
            }
            with(holder.itemBinding.checkBox) {
                tag = position
                setOnCheckedChangeListener(null)
                isChecked = isSelected
                setOnCheckedChangeListener(this@MyAdapter)
            }
        }

        // Return the size of your dataSet (invoked by the layout manager)
        override fun getItemCount() = dataSet.size

        override fun getItemViewType(position: Int) = if (position == viewModel.headerLine) 0 else 1
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        viewModel.mapping = (0 until nrOfColumns).map {
            (binding.headerLine.getChildAt(it + 1) as Spinner).selectedItemPosition
        }
            .toIntArray()
    }

    @Deprecated("Deprecated in Java")
    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.csv_import, menu)
    }

    @Deprecated("Deprecated in Java")
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.IMPORT_COMMAND) {
            val columnToFieldMap = IntArray(nrOfColumns)
            val header = viewModel.headerLine.takeIf { it > -1 }?.let { dataSet[it] }
            for (i in 0 until nrOfColumns) {
                val position =
                    (binding.headerLine.getChildAt(i + 1) as Spinner).selectedItemPosition
                columnToFieldMap[i] = fields[position].first
                if (position > 0) {
                    header?.let {
                        if (header.isSet(i)) {
                            try {
                                header2FieldMap.put(
                                    Utils.normalize(header[i]),
                                    fields[position].second
                                )
                            } catch (e: JSONException) {
                                CrashHandler.report(e)
                            }
                        }
                    }
                }
            }
            if (validateMapping(columnToFieldMap)) {
                prefHandler.putString(
                    PrefKey.CSV_IMPORT_HEADER_TO_FIELD_MAP,
                    header2FieldMap.toString()
                )
                val selectedData = dataSet.filterIndexed { index, _ -> viewModel.isSelected(index) }
                (activity as? CsvImportActivity)?.importData(
                    selectedData,
                    columnToFieldMap,
                    dataSet.size - selectedData.size
                )
            }
        }
        return super.onOptionsItemSelected(item)
    }

    /**
     * Check mapping constraints:<br></br>
     *
     *  * No field mapped more than once
     *  * Subcategory cannot be mapped without category
     *  * One of amount, income or expense must be mapped.
     *
     * @param columnToFieldMap
     */
    private fun validateMapping(columnToFieldMap: IntArray): Boolean {
        val foundFields = SparseBooleanArray()
        val activity = requireActivity() as ProtectedFragmentActivity
        for (field in columnToFieldMap) {
            if (field != R.string.discard) {
                if (foundFields[field, false]) {
                    activity.showSnackBar(
                        getString(
                            R.string.csv_import_field_mapped_more_than_once,
                            getString(field)
                        )
                    )
                    return false
                }
                foundFields.put(field, true)
            }
        }
        if (foundFields[R.string.subcategory, false] && !foundFields[R.string.category, false]) {
            activity.showSnackBar(R.string.csv_import_subcategory_requires_category)
            return false
        }
        if (!(foundFields[R.string.amount, false] ||
                    foundFields[R.string.expense, false] ||
                    foundFields[R.string.income, false])
        ) {
            activity.showSnackBar(R.string.csv_import_no_mapping_found_for_amount)
            return false
        }
        return true
    }

    private fun firstSelectedRow() = dataSet.indices.find { viewModel.isSelected(it) } ?: -1

    companion object {
        fun newInstance() = CsvImportDataFragment()
    }
}